[docs] usecase: opentelemetry (#3439)
feldera/feldera: docs/use_cases/otel/preprocessing.md

# Ingress

To push data into Feldera, we configure the OpenTelemetry (OTel) Collector to send data to Feldera via HTTP.

#### OTel Collector Configuration:

> ```yml

Reviewer: this yml file is fed to?

exporters:

  # Send traces to the Feldera pipeline.
  # Currently only uncompressed JSON is supported.
  otlphttp/feldera_traces:
     traces_endpoint: http://feldera:8080/v0/pipelines/otel/ingress/otel_traces?format=json&update_format=raw
     encoding: json
     compression: none

  otlphttp/feldera_logs:
     logs_endpoint: http://feldera:8080/v0/pipelines/otel/ingress/otel_logs?format=json&update_format=raw
     encoding: json
     compression: none

  otlphttp/feldera_metrics:
     metrics_endpoint: http://feldera:8080/v0/pipelines/otel/ingress/otel_metrics?format=json&update_format=raw
     encoding: json
     compression: none

service:
   pipelines:
     traces:
       exporters: [spanmetrics, otlphttp/feldera_traces]

     logs:
       receivers: [otlp]
       exporters: [otlphttp/feldera_logs]

     metrics:
       receivers: [hostmetrics, docker_stats, httpcheck/frontend-proxy, otlp, prometheus, redis, spanmetrics]
       exporters: [otlphttp/feldera_metrics]
```

Notice that we have the query params: `format=json&update_format=raw`
- `format=json`: Specifies that the input data is JSON-formatted.

> - `update_format=raw`: Indicates that the request body contains raw data, which is processed as an insertion into the table.

Reviewer: who is processing it as an insertion in a table?
is it the Feldera adapter?

## Preprocessing

> Next we perform a series of `UNNEST` operations to extract individual rows of metrics, logs and traces from a nested arrays of them.

Reviewer: you are now shifting to a different box in the dataflow. Say it.

Reviewer: this SQL program is to be concatenated with the type declarations shown before.
so add a comment in both programs saying this.

Reviewer: explain the high-level point: the data contains arrays of fields, and you want to convert it into traditional database table rows, for easier processing in SQL.

#### Step 1: Extract Resource-Level Data

```sql
-- (ResouceMetrics[N]) -> (Resource, ScopeMetrics[N])
CREATE LOCAL VIEW rsMetrics AS SELECT resource, scopeMetrics
FROM otel_metrics, UNNEST(resourceMetrics) as t (resource, scopeMetrics);

-- (ResouceSpans[N]) -> (Resource, ScopeSpans[N])
CREATE LOCAL VIEW rsSpans AS SELECT resource, scopeSpans
FROM otel_traces, UNNEST(resourceSpans) as t (resource, scopeSpans);

-- (ResouceLogs[N]) -> (Resource, ScopeLogs[N])
CREATE LOCAL VIEW rsLogs AS SELECT resource, scopeLogs
FROM otel_logs, UNNEST(resourceLogs) as t (resource, scopeLogs);
```

> Here we `UNNEST`  arrays of `ResourceSpans`, `ResourceMetrics` and `ResourceLogs` into separate rows.

Reviewer: Some observability people may not be familiar with SQL, so it may pay off to have a simple running example with a concrete event that shows how it will look in each view.

#### Step 2: Extract Scope-Level Data

```sql
-- (ScopeMetrics[N]) -> (ScopeMetrics) x N
CREATE LOCAL VIEW metrics_array AS
SELECT metrics From rsMetrics, UNNEST(rsMetrics.scopeMetrics) as t(_, metrics);

-- (ScopeLogs[N]) -> (ScopeLogs) x N
CREATE LOCAL VIEW logs_array AS
SELECT logs FROM rsLogs, UNNEST(rsLogs.scopeLogs) as t(_, logs);

-- (ScopeSpans[N]) -> (ScopeSpans) x N
CREATE LOCAL VIEW spans_array AS
SELECT spans FROM rsSpans, UNNEST(rsSpans.scopeSpans) as t(_, spans);
```

Similarly, we `UNNEST` the `ScopeMetrics`, `ScopeLogs` and `ScopeSpans` into separate rows.

#### Step 3: Extract Final Metrics and Logs

```sql
-- (Metrics[N]) -> (_, Metric) x N
CREATE MATERIALIZED VIEW metrics AS
SELECT
	name,
	description,
	unit,
	data,
	metadata
FROM metrics_array, UNNEST(metrics_array.metrics);

-- (Logs[N]) -> (_, Logs) x N
CREATE MATERIALIZED VIEW logs AS
SELECT
	attributes,
	timeUnixNano,
	observedTimeUnixNano,
	severityNumber,
	severityText,
	flags,
	traceId,
	spanId,
	eventName,
	body
FROM logs_array, UNNEST(logs_array.logs);
```

> Finally, we extract individual `Metric` and `LogRecords` to extract individual records.

Reviewer: we have extracted.
This could be a better comment in the code than the existing one.

#### Step 4: Process Spans and Add Derived Fields

For spans, we not only extract individual records but also compute useful derived fields:
- `elapsedTimeMillis`: The duration of the span in milliseconds.

> - `eventTime`: Timestamp for when the span started. Note that in Feldera, the `TIMESTAMP` type doesn't include time zone information.

Reviewer: timestamp indicating when the span started.
The note can be a footnote, we will add support for timezone.

> # Ingress

Reviewer 2: ```suggestion
# Sending OTel data to Feldera
```

> To push data into Feldera, we configure the OpenTelemetry (OTel) Collector to send data to Feldera via HTTP.

Reviewer 2: Reference to HTTP ingest docs

#### OTel Collector Configuration:

> Here we configure OTel Collector to send this data to Feldera as JSON via an HTTP POST request without any compression.

Reviewer 2: ```suggestion
Here we configure OTel Collector to send logs, traces, and metrics to Feldera as JSON via an HTTP POST request without any compression.
```

![OTel Collector to Feldera Table](otel-feldera-table.png)

Configure your OTel Collector as follows:

```yml
# src/otel-collector/otelcol-config-extras.yml in OpenTelemetry Demo

exporters:

  # Send traces to the Feldera pipeline.
  # Currently only uncompressed JSON is supported.
  otlphttp/feldera:
     logs_endpoint: http://feldera:8080/v0/pipelines/otel/ingress/otel_logs?format=json&update_format=raw
     traces_endpoint: http://feldera:8080/v0/pipelines/otel/ingress/otel_traces?format=json&update_format=raw
     metrics_endpoint: http://feldera:8080/v0/pipelines/otel/ingress/otel_metrics?format=json&update_format=raw
     encoding: json
     compression: none

service:
   pipelines:
     traces:
       receivers: [otlp]
       processors: [transform, batch]
       exporters: [spanmetrics, otlphttp/feldera, debug, otlp]

     logs:
       receivers: [otlp]
       processors: [batch]
       exporters: [otlphttp/feldera, debug]

     metrics:
       receivers: [hostmetrics, docker_stats, httpcheck/frontend-proxy, otlp, prometheus, redis, spanmetrics]
       processors: [batch]
       exporters: [otlphttp/feldera, debug]
```

> Notice that we have the query parameters: `format=json&update_format=raw`

Reviewer 2: add references to docs
